File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.go
func lengthOfLongestSubstring(s string) int {
    //test cases: 
    //  "dvdf" 
    longest := 0
    startIndex := 0 
    endIndex := 0
    collection := make(map[rune]int)
    for pos, char := range(s) {
        if val, ok := collection[char]; ok{
            //fmt.Println(collection)
            //fmt.Println(startIndex, endIndex)
            distance := endIndex - startIndex 
            //fmt.Println(distance)
            if distance > longest{
                //fmt.Println(longest , "------------>" , distance)
                longest = distance
            }
            
            newStart := val + 1
            if newStart > startIndex{
                startIndex = newStart
            }
            
            delete(collection,char)
        }
        
        collection[char] = pos
        endIndex++
    }
    distance := endIndex - startIndex 
    if distance > longest{
        //fmt.Println(longest , "+++++++>" , distance)
        longest = distance
    }
    return longest
}

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    # test cases:
    #   "bvvf"
    #   "au"
    #   "abba"
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0:
            return 0
        
        dic_substring = {}
        index_substring = 0
        index_start = 0
        longest = 1
        for char in s:
            if char in dic_substring:
                            #index_substring -1 - index_start +1
                distance = index_substring - index_start
                
                if distance > longest:
                    longest = distance
                    #print("start: {:n} , stop: {:n}, longest: {:n}".format(index_start, index_substring, longest))
                
                index_start = max(index_start, dic_substring[char] + 1)
                #print("index_start: ",index_start)
                del dic_substring[char]
                    
            dic_substring[char] = index_substring
            index_substring +=1
            #print(dic_substring)
            
        distance = index_substring - index_start
        if distance > longest:
            longest = distance
            #print("start: {:n} , stop: {:n}, longest: {:n}".format(index_start, index_substring, longest))
            
        return longest